pkg/registry/core/shoot/storage/admin_kubeconfig.go

```diff
@@ -0,0 +1,184 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package storage
+
+import (
+	"context"
+	"fmt"
+	"net/url"
+	"time"
+
+	authenticationapi "github.com/gardener/gardener/pkg/apis/authentication"
+	authenticationapiv1alpha1 "github.com/gardener/gardener/pkg/apis/authentication/v1alpha1"
+	authenticationvalidation "github.com/gardener/gardener/pkg/apis/authentication/validation"
+	"github.com/gardener/gardener/pkg/apis/core"
+	gardencorev1alpha1 "github.com/gardener/gardener/pkg/apis/core/v1alpha1"
+	v1beta1constants "github.com/gardener/gardener/pkg/apis/core/v1beta1/constants"
+	"github.com/gardener/gardener/pkg/client/kubernetes"
+	"github.com/gardener/gardener/pkg/utils/infodata"
+	"github.com/gardener/gardener/pkg/utils/secrets"
+
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apiserver/pkg/authentication/user"
+	genericapirequest "k8s.io/apiserver/pkg/endpoints/request"
+	"k8s.io/apiserver/pkg/registry/rest"
+)
+
+// AdminKubeconfigREST implements a RESTStorage for shoots/adminkubeconfig.
+type AdminKubeconfigREST struct {
+	shootStateStorage    getter
+	shootStorage         getter
+	now                  func() time.Time
+	maxExpirationSeconds int64
+}
+
+var (
+	_ = rest.NamedCreater(&AdminKubeconfigREST{})
+	_ = rest.GroupVersionKindProvider(&AdminKubeconfigREST{})
+
+	gvk = schema.GroupVersionKind{
+		Group:   authenticationapiv1alpha1.SchemeGroupVersion.Group,
+		Version: authenticationapiv1alpha1.SchemeGroupVersion.Version,
+		Kind:    "AdminKubeconfigRequest",
+	}
+)
+
+// New returns and instance of AdminKubeconfigRequest
+func (r *AdminKubeconfigREST) New() runtime.Object {
+	return &authenticationapi.AdminKubeconfigRequest{}
+}
+
+// Create returns a AdminKubeconfigRequest with kubeconfig based on
+// - shoot's advertised addresses
+// - shoot's certificate authority
+// - user making the request
+func (r *AdminKubeconfigREST) Create(ctx context.Context, name string, obj runtime.Object, createValidation rest.ValidateObjectFunc, _ *metav1.CreateOptions) (runtime.Object, error) {
+	if createValidation != nil {
+		if err := createValidation(ctx, obj.DeepCopyObject()); err != nil {
+			return nil, err
+		}
+	}
+
+	out := obj.(*authenticationapi.AdminKubeconfigRequest)
+
+	if errs := authenticationvalidation.ValidateAdminKubeconfigRequest(out); len(errs) != 0 {
+		return nil, errors.NewInvalid(gvk.GroupKind(), "", errs)
+	}
+
+	userInfo, ok := genericapirequest.UserFrom(ctx)
+	if !ok {
+		return nil, errors.NewBadRequest("no user in context")
+	}
+
+	shootStateObj, err := r.shootStateStorage.Get(ctx, name, &metav1.GetOptions{})
+	if err != nil {
+		return nil, err
+	}
+
+	shootState := &gardencorev1alpha1.ShootState{}
+
+	err = kubernetes.GardenScheme.Convert(shootStateObj, shootState, nil)
+	if err != nil {
+		return nil, err
+	}
```

Comment:
Why is this needed? Can't you work with the internal type only?

Reply:
Due to https://github.com/gardener/gardener/blob/39929464663589fb6a0b234153d4fb998c36ed87/pkg/utils/infodata/infodata.go#L61

It accepts only https://github.com/gardener/gardener/blob/95b83f3a1f3fe0c8c41dbf8193658a2d5295e4c3/pkg/apis/core/v1alpha1/helper/shootstate_list.go#L66-L67

Reply:
👍🏻 Makes sense, thank you.